code/modules/mob/living/simple_animal/hostile/space_dragon.dm

```diff
@@ -0,0 +1,727 @@
+	var/carp_interval = 60
+	/// The time since an extra carp was added to the ghost role spawning pool.
+	var/last_carp_inc = 0
+	/// A list of all the ckeys which have used this carp rift to spawn in as carps.
+	var/list/ckey_list = list()
+
+/obj/structure/carp_rift/Initialize(mapload)
+	. = ..()
+	START_PROCESSING(SSobj, src)
+
+/obj/structure/carp_rift/examine(mob/user)
+	. = ..()
+	if(time_charged < max_charge)
+		. += "<span class='notice'>It seems to be [(time_charged / max_charge) * 100]% charged.</span>"
+	else
+		. += "<span class='warning'>This one is fully charged.  In this state, it is poised to transport a much larger amount of carp than normal.</span>"
+
+	if(isobserver(user))
+		. += "<span class='notice'>It has [carp_stored] carp available to spawn as.</span>"
+
+/obj/structure/carp_rift/play_attack_sound(damage_amount, damage_type = BRUTE, damage_flag = 0)
+	playsound(src, 'sound/magic/lightningshock.ogg', 50, TRUE)
+
+/obj/structure/carp_rift/Destroy()
+	STOP_PROCESSING(SSobj, src)
+	if(time_charged != max_charge + 1)
+		dragon?.destroy_rifts()
+		if(dragon)
+			to_chat(dragon, "<span class='boldwarning'>A rift has been destroyed! You have failed, and find yourself weakened.</span>")
+	return ..()
+
+/obj/structure/carp_rift/process(delta_time)
+	// Heal carp on our loc.
+	for(var/mob/living/simple_animal/hostile/hostilehere in loc)
+		if("carp" in hostilehere.faction)
+			hostilehere.adjustHealth(-10)
+			var/obj/effect/temp_visual/heal/H = new /obj/effect/temp_visual/heal(get_turf(hostilehere))
+			H.color = "#0000FF"
+
+	// If we're fully charged, just start mass spawning carp and move around.
+	if(charge_state == CHARGE_COMPLETED)
+		if(DT_PROB(1.25, delta_time))
+			new /mob/living/simple_animal/hostile/carp(loc)
+		if(DT_PROB(1.5, delta_time))
+			var/rand_dir = pick(GLOB.cardinals)
+			Move(get_step(src, rand_dir), rand_dir)
+		return
+
+	// Increase time trackers and check for any updated states.
+	time_charged = min(time_charged + delta_time, max_charge)
+	last_carp_inc += delta_time
+	update_check()
+
+/obj/structure/carp_rift/attack_ghost(mob/user)
+	. = ..()
+	summon_carp(user)
+
+/**
+ * Does a series of checks based on the portal's status.
+ *
+ * Performs a number of checks based on the current charge of the portal, and triggers various effects accordingly.
+ * If the current charge is a multiple of carp_interval, add an extra carp spawn.
+ * If we're halfway charged, announce to the crew our location in a CENTCOM announcement.
+ * If we're fully charged, tell the crew we are, change our color to yellow, become invulnerable, and give Space Dragon the ability to make another rift, if he hasn't summoned 3 total.
+ */
+/obj/structure/carp_rift/proc/update_check()
+	// If the rift is fully charged, there's nothing to do here anymore.
+	if(charge_state == CHARGE_COMPLETED)
+		return
+
+	// Can we increase the carp spawn pool size?
+	if(last_carp_inc >= carp_interval)
```

Comment:
Use a world.time based timer

Reply:
This actually works better by adding delta time as it currently does, the number never gets larger than `carp_interval` and using `addtimers` doesn't really work well since the interval is dynamic. If the interval changes, this needs to respond instantly, `world.time` operates on game ticks and will not work well during lag.

I fixed the other one, but this one should stay as it is